### DISCUSSION TITLE
How to manually shrink a volume?

### DISCUSSION THREAD

**Discussion by iosifnicolae2** [longhorn/longhorn#4178] [Show and tell]
### **VERY IMPORTANT!**
**1. Make sure to use persistentVolumeReclaimPolicy: Retain otherwise the PV will be deleted when deleting the PVC**
**2.Not recommended for HA production deployments**
**3. The below procedure is very custom for our workflow and might not work for you**

Here is the process we've followed to shrink some volumes:
1. Scale all deployments using the volumes to 0 (stop all operators)
1. Delete all Persistent Volume Claims
1. Scale down the volume size in your K8S manifest
1. Scale all deployments back to normal so another PVC's are created / start K8S operator
1. Scale all deployments using the volumes to 0 (stop all operators)
1. Attach the two PVC's to the same node using the Longhorn interface (do not check the Maintenance mode)
```bash
mkdir /source /target
umount /source /target
mount /dev/longhorn/pvc-SOURCE_PVC_HERE /source
mount /dev/longhorn/pvc-TARGET_PVC_HERE /target
rm -r /target/*
rsync -a /source/* /target/
umount /source /target
```
7. Detach all volumes from Longhorn interface
7. Scale all deployments back to normal / start K8S operator

**Comment by jemand771** [comment]
Just as a minor addition that might make it easier depending on the actual use case:
I really enjoy using https://github.com/utkuozdemir/pv-migrate for moving files from one pv to another, even across namespaces or different storage classes (e.g. when migrating to longhorn)

The workflow for using it is similar to yours, just without going too deep into the internals:
1. scale down workloads
2. delete pvc but keep the pv
3. create a new pvc (with a different, temporary name) that's manually bound to the old pv
4. create new pvc (auto-creates longhorn volume) that will be used in the end (desired new size)
5. run pv-migrate (example: `kubectl pv-migrate migrate -n minio -N minio -s mnt2 data-minio-0-old data-minio-0`; see the command's `--help` output for more info on which transfer strategies (`-s`) exist)
6. scale up deployment
7. clean up temporary pvc and old pv

because I'm lazy I usually do one extra step to save myself from manually binding the old pv to a pvc:
* create a temporary pvc (same manifest as old, just with a different name)
* migrate from old production volume to tmp
* delete old production pv and pvc
* create new production pv and pvc (same as old prod one but smaller)
* migrate from tmp to new volume
* delete tmp volume

(you might also need to do that if your reclaimpolicy is delete, although there might be other ways around that)
